src/databricks/labs/ucx/framework/owners.py

```diff
@@ -0,0 +1,204 @@
+import functools
+import logging
+from abc import ABC, abstractmethod
+from collections.abc import Callable, Iterable, Sequence
+from functools import cached_property
+from typing import ClassVar, Generic, Protocol, TypeVar, final
+
+from databricks.sdk import WorkspaceClient
+from databricks.sdk.errors import NotFound
+from databricks.sdk.service.iam import User
+
+logger = logging.getLogger(__name__)
+
+
+class DataclassInstance(Protocol):
+    __dataclass_fields__: ClassVar[dict]
+
+
+Record = TypeVar("Record")
+
+
+class AdministratorFinder(ABC):
+    def __init__(self, ws: WorkspaceClient) -> None:
+        self._ws = ws
+
+    @abstractmethod
+    def find_admin_users(self) -> Iterable[User]:
+        """Locate active admin users."""
+        raise NotImplementedError()
+
+
+class WorkspaceAdministratorFinder(AdministratorFinder):
+    """Locate the users that are in the 'admin' workspace group for a given workspace."""
+
+    @staticmethod
+    def _member_of_group_named(user: User, group_name: str) -> bool:
+        """Determine whether a user belongs to a group with the given name or not."""
+        return user.groups is not None and any(g.display == group_name for g in user.groups)
+
+    @staticmethod
+    def _member_of_group(user: User, group_id: str) -> bool:
+        """Determine whether a user belongs to a group with the given identifier or not."""
+        return user.groups is not None and any(g.value == group_id for g in user.groups)
+
+    def _is_active_admin(self, user: User) -> bool:
+        """Determine if a user is an active administrator."""
+        return bool(user.active) and self._member_of_group_named(user, "admins")
+
+    def _filter_workspace_groups(self, identifiers: Iterable[str]) -> Iterable[str]:
+        """Limit a set of identifiers to those that are workspace groups."""
+        seen = set()
+        for group_id in identifiers:
+            if group_id in seen:
+                continue
+            seen.add(group_id)
```

Comment:
```suggestion
        for group_id in set(identifiers):
```
you want to process each group only once, even if it's supplied twice, right?

Reply:
Indeed, this was clumsy. I'm simplified it.